Resolve errors in perfkernel for Windows (#15753)

jspark1105 opened the pull request:
Summary:
1. MSVC only has _mm_prefetch(const char*, int). Fixed in both python codegen and C++ files.
2. uint32_t in "cvtsh_ss_bugfix.h" requires "#include <cstdint>".
3. Some files use gflags headers. Add dependency via c10.
4. Isolate arch flags with interface library and private compile options.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15753

Differential Revision: D13636233

dskhudia commented on caffe2/perfkernels/CMakeLists.txt:
```diff
@@ -13,37 +13,63 @@ set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS} ${common_srcs})
+    target_compile_options(Caffe2_perfkernels_avx
+        PRIVATE "/arch:AVX")
+    target_compile_options(Caffe2_perfkernels_avx2
+        PRIVATE "/arch:AVX2")
```
I am not sure how AVX and AVX2 options are defined for MSVC but do we need to add corresponding options for -mfma, mf16c etc.? Are they subsumed in AVX2 for MSVC?

jspark1105 replied:
Based on what @xkszltl's explanation, unlike -mavx* options in gcc/clang, arch:AVX* options in MSVC only control compiler's auto generation of avx* instructions and defines a few preprocessor symbols. Intrinsics are available regardless. Not sure if __F16C__ and __FMA__ will be defined by /arch:AVX and /arch:AVX2 , respectively. If not, we may want to define them here.

xkszltl replied:
__FMA__ is not defined as a warning from cpuinfo (machine has FMA but not build with it) popped out.

jspark1105 replied:
Thanks for the information. I'll add D__FMA__ to avx2 and D__F16C__ to avx.

dskhudia commented on caffe2/perfkernels/CMakeLists.txt:
```diff
@@ -13,37 +13,63 @@ set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS} ${common_srcs})
+    add_dependencies(Caffe2_perfkernels_avx512 Caffe2_PROTO)
+    target_link_libraries(Caffe2_perfkernels_avx512 PRIVATE c10)
+    if (MSVC)
+      target_compile_options(Caffe2_perfkernels_avx512
```
Do we need avx2 and avx (and fma (see above)) options here as well?

xkszltl committed:
Resolve errors in perfkernel for Windows (#16031)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16031

1. MSVC only has _mm_prefetch(const char*, int). Fixed in both python codegen and C++ files.
2. uint32_t in "cvtsh_ss_bugfix.h" requires "#include <cstdint>".
3. Some files use gflags headers. Add dependency via c10.
4. Isolate arch flags with interface library and private compile options.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15753

Differential Revision: D13636233

fbshipit-source-id: 3e2407db0f4c2db57ed7284618a86c459832907e
